src/csexp_rpc/csexp_rpc.ml

```diff
@@ -52,67 +49,71 @@ module Session = struct
     | None -> "EOF"
     | Some sexps -> String.concat ~sep:" " (List.map ~f:Sexp.to_string sexps)
 
+  let close t =
+    match t.state with
+    | Closed -> ()
+    | Open { in_channel; out_channel; reader; writer } ->
+      Worker.stop reader;
+      Worker.stop writer;
+      close_out_noerr out_channel;
+      close_in_noerr in_channel;
```

Comment:
Since your in and out channels are backed by the same fd, I think we have a double-close here, which is bad. (if there are multiple threads involved, then it's possible for the second `close` to hit an unrelated file descriptor)

A solution that should work is to use `Unix.dup` before creating the channels so they are backed by different fds.

Note, however, that (from what I remember) the automatic socket tear-down only happens when the last fd to that socket is closed, so "closing" the write end alone won't be sufficient to send EOF. This is probably fine if you're closing both read and write end together, but I have a vague memory that using shutdown is still recommended. (I seem to be repeating myself)

Reply:
I revived some old code to distinguish the two cases after all